plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,186 @@
+ With the applications:
+ @
+   (force^|Tn| (... (force^|T3| ((force^|T2| ((force^|T1| original) X1)) X2)) X3 ...)) Xn
+ @
+ After inlining @original@ we get:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((force^|T1|
+             (delay^|T1|
+               (\X1 ->
+                  delay^|T2| (\X2 ->
+                    delay^|T3| (\X3 ->
+                      ... ->
+                        delay^|Tn| (\Xn -> body)))))
+            ) X1)) X2)) X3 ...)) Xn
+ @
+ In the end, after applying the "UntypedPlutusCore.Transform.ForceDelay" optimisation:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((\X1 ->
+             delay^|T2| (\X2 ->
+               delay^|T3| (\X3 ->
+                 ... ->
+                   delay^|Tn| (\Xn -> body)))))
+        X1) X2)) X3 ...)) Xn
+ @
+ Notice that the next two reduction steps (applying @X1@ and reducing @force (delay ...)@)
+ produce an equivalent term to applying the transformation and then the reduction rule
+ for application.
+ This is easy to check, so we continue by showing what the "optimised term" looks like:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((\X1 -> \X2 ->
+           delay^|T3| (\X3 ->
+             ... ->
+               delay^|Tn| (\Xn -> body))
+        X1) X2)) X3 ...)) Xn
+ @
+ The term can be optimised further by "erasing" the @force^|T3|@ and @delay^|T3|@ pair,
+ and so on until @Tn@.
+-}
+{-# LANGUAGE LambdaCase #-}
+module UntypedPlutusCore.Transform.ForceApply
+    ( forceApply
+    ) where
+
+import UntypedPlutusCore.Core
+
+import Control.Lens (transformOf)
+import Control.Monad (guard)
+import Data.Foldable (foldl')
+import Data.Maybe (fromMaybe)
+
+{- | Traverses the term, for each node applying the optimisation
+ detailed above. For implementation details see 'optimisationProcedure'.
+
+ Note: this optimisation should be applied after
+ 'UntypedPlutusCore.Transform.ForceDelay.forceDelay' for reasons mentioned
+ above.
+-}
+forceApply :: Term name uni fun a -> Term name uni fun a
+forceApply = transformOf termSubterms processTerm
+
+{- | Checks whether the term is of the right form, and "pushes"
+ the 'Force' down into the underlying lambda abstractions.
+-}
+processTerm :: Term name uni fun a -> Term name uni fun a
+processTerm = \case
+    original@(Force _ subTerm) ->
+        fromMaybe original (optimisationProcedure subTerm)
+    t -> t
+
+{- | Converts the subterm of a 'Force' into specialised types for representing
+ multiple applications on top of multiple abstractions. Checks whether the lambda
+ will eventually get "fully reduced" and applies the optimisation.
+ Returns 'Nothing' if the optimisation cannot be applied.
+-}
+optimisationProcedure :: Term name uni fun a -> Maybe (Term name uni fun a)
+optimisationProcedure term = do
+    asMultiApply <- toMultiApply term
+    innerMultiAbs <- toMultiAbs . toApply $ asMultiApply
```

Comment:
It's rather weird that we have force-delay as a separate optimization given that `toMultiApply` detecting no applications would give pretty much precisely force-delay, wouldn't it?

Reply:
I would agree with @michaelpj on this. In my experience, simplifiers can get _very_ tricky to understand if you're not careful. Keeping concerns as separate as possible is one way to make sure that each pass does it's own explicitly defined thing without any hidden complexity.